### PR TITLE
Ah/testing

### PR DESCRIPTION
**Pull request by jibben**
guys pretty printing!!!!

### CODE REVIEW

**Comment by Howon** [comment]
Maybe this should be merged into Ast or renamed debug?

File: src/prettyprint.ml [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+(* Print Expressions BRUG*)
+
+let rec string_of_expr = function
+    Binop(e1, o, e2)	        -> "(" ^
+                                  (string_of_expr e1) ^ " " ^
+                                  (string_of_binop o) ^ " " ^
+							                    (string_of_expr e2)
+                                  ^ ")"
+  | Unop(o, e)                -> (string_of_expr e) ^ (string_of_uop o)
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: should be reversed:

`(string_of_uop o) ^ (string_of_expr e)`

File: src/prettyprint.ml [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  | Id(s)                     -> s
+  | Assign(v, e)              -> v ^ " = " ^ string_of_expr e
+  | Access(v, e)              -> v ^ "[" ^ string_of_expr e ^ "]"
+  | Int_Lit(i)			          -> string_of_int i
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: plz fix spacing

File: src/prettyprint.ml [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  | Double_Lit(f)             -> string_of_float f
+  | Char_Lit(c)               -> Char.escaped c
+  | String_Lit(s)		          -> s
```
Comment (vlad-scherbich) [inline src/prettyprint.ml]: spacing

File: src/prettyprint.ml [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  | Expr(expr)                    -> string_of_expr expr ^ ";"
+  | Return(expr)                  -> "return " ^ string_of_expr expr ^ ";"
+  | Mut(v, t)                     -> "mut " ^ string_of_types t ^ " " ^
+                                      v ^ ";"
+  | Mutdecl(v, t, e)              -> "mut " ^  string_of_types t ^ " " ^ v ^
+                                      string_of_expr e ^";"
+  | Vdecl(v, t, e)                -> string_of_types t ^ " " ^ v ^
+                                      string_of_expr e ^ ";"
+  | If(e, s1, s2)                 -> "if (" ^ string_of_expr e ^ ") {\n" ^
+                                      string_of_stmts s1 ^ "\n} else {\n" ^
```
Comment (vlad-scherbich) [inline src/prettyprint.ml:136]: should `else` be here? what if you just have an `if` branch?

File: src/prettyprint.ml [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+and string_of_lambda lambda =
+  "(" ^ string_of_formals lambda.l_formals ^ ") => " ^ string_of_types
+    lambda.l_return_t ^ " = {\n" ^ string_of_stmts lambda.l_body ^ "\n}"
+
+
+(* Print Patterns BRUG*)
+
+let string_of_pattern p =
+  "| " ^ p.p_message_id ^ "(" ^ string_of_formals p.p_message_formals
+    ^ ") => {\n        " ^ String.concat "\n        "
```
Comment (vlad-scherbich) [inline src/prettyprint.ml:168]: do we need the whitespace after both `\n`s?

File: src/prettyprint.ml [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    (List.map string_of_stmt p.p_stmts) ^ "\n      }"
```
Comment (vlad-scherbich) [inline src/prettyprint.ml:169]: here too ^^

File: src/prettyprint.ml [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+  "def " ^ func.f_name ^ "(" ^ string_of_formals func.f_formals ^ ") => " ^
+    string_of_types func.f_return_t ^ " = {\n" ^
+    string_of_stmts func.f_body ^ "\n}"
+
+
+(* Print Actor BRUG*)
+
+let string_of_actor actor =
+  "actor " ^ actor.a_name ^ "(" ^ string_of_formals actor.a_formals ^
+  ") {\n  " ^ string_of_stmts actor.a_body ^ "\n\n  " ^
```
Comment (vlad-scherbich) [inline src/prettyprint.ml:184]: whitespace after `\n`s here?

**Review by vlad-scherbich — APPROVED** [review]

File: src/prettyprint.ml (hunk 4 above)
Comment (jibben) [reply src/prettyprint.ml:136]: Under the hood, every else has an else. I could add some logic to only print the else if the length of s2 is greater than 0, but I don't think that's worth it

Comment (vlad-scherbich) [reply src/prettyprint.ml:136]: OK, less merge dis!

**Pull request merged**